pynxtools/dataconverter/readers/stm/stm_file_parser.py

```diff
@@ -0,0 +1,383 @@
+                        axes_data.append(np.linspace(*coor_info[0][0:2], x_cor_len))
+                        axes_data.append(np.linspace(*coor_info[1][0:2], y_cor_len))
+                    axes_units.append(coor_info[0][2])
+                    template[temp_data_field] = scan_dt_arr
+                else:
+                    # to clean up nxdata_grp and field_name from previous loop
+                    nxdata_grp = ''
+                    field_name = ''
+
+        def fill_out_NXdata_group():
+            """To fill out NXdata which is root for all data fields and attributes for NXdata.
+            This function fills template with first level of descendent fields and attributes
+            of NXdata but not the fields and attributes under child of NXdata.
+            """
+            if nxdata_grp:
+                auxiliary_signals_attr = f"{nxdata_grp}/@auxiliary_signals"
+                axes = f"{nxdata_grp}/@axes"
+                signal_attr = f"{nxdata_grp}/@signal"
+                template[auxiliary_signals_attr] = []
+                template[axes] = axes_name
+                for ind, data_field_nm in enumerate(signals):
+
+                    if ind == 0:
+                        template[signal_attr] = data_field_nm
+                    else:
+                        template[auxiliary_signals_attr].append(data_field_nm)
+                for axis, axis_data in zip(axes_name, axes_data):
+                    template[f"{nxdata_grp}/{axis}"] = axis_data
+
+        def find_nxdata_group_and_name(key):
+            """Find data group name from a data path in file.
+            E.g. 'Z', 'LI_Demod_2_X' from /Z/forward and /LI_Demod_2_X/forward
+            Note: Create a function in stm_helper.py to unit scale such as nm, micrometer
+            """
+            tmp_key = key.split('/', 1)[1]
+            grp_name, data_field_name = tmp_key.split('/', 1)
+            grp_name = grp_name.upper()
+            return grp_name, data_field_name
+
+        for _, dt_path_list in sub_config_dict.items():
+            signals = []
+            axes_name = ['x', 'y']
+            axes_units = []
+            axes_data = []
+            # The following functions can be thought as unpacked function body here.
+            if not dt_path_list:
+                continue
+            indivisual_DATA_field()
+            fill_out_NXdata_group()
+
+    # pylint: disable=too-many-locals
+    def get_dimension_info(self, config_dict, data_dict):
+        """
+        Extract dimension info from scanfield.
+
+            ../ENVIRONMENT[environment]/scan_control/positioner/scanfield"
+            The scanfield has four parts starting point of (x, y) co-ordinate
+            length on (x, y)-dimenstion and one last unknown values.
+        """
+        scanfield: str = ''
+        for key, val in config_dict.items():
+            if ("/ENTRY[entry]/INSTRUMENT[instrument]/ENVIRONMENT[environment]/"
```

Comment:
But this line for instance shows that the left to right mapping (domain file format to NeXus mapping) is maybe quite stiff still because what if we have an ENTRY[entry2] can this be dealt with?
I think for a first round of the parser this is all good...